src/dragonball/src/device_manager/vfio_dev_mgr/mod.rs

```diff
@@ -0,0 +1,805 @@
+    NoResource,
+
+    /// Cannot perform the requested operation after booting the microVM
+    #[error("update operation is not allowed after boot")]
+    UpdateNotAllowedPostBoot,
+
+    /// Failed to create kvm device
+    #[error("failed to create kvm device: {0:?}")]
+    CreateKvmDevice(#[source] vmm_sys_util::errno::Error),
+
+    /// Failed to restore vfio mlock count
+    #[error("failure while restoring vfio mlock count: {0:?}")]
+    RestoreMlockCount(#[source] std::io::Error),
+
+    /// Failure in device manager while managing VFIO device
+    #[error("failure in device manager while managing VFIO device, {0:?}")]
+    VfioDeviceMgr(#[source] DeviceMgrError),
+
+    /// Failure in VFIO IOCTL subsystem.
+    #[error("failure while configuring VFIO device, {0:?}")]
+    VfioIoctlError(#[source] vfio_ioctls::VfioError),
+
+    /// Failure in VFIO PCI subsystem.
+    #[error("failure while managing PCI VFIO device: {0:?}")]
+    VfioPciError(#[source] dbs_pci::VfioPciError),
+
+    /// Failure in PCI subsystem.
+    #[error("PCI subsystem failed to manage the device: {0:?}")]
+    PciError(#[source] dbs_pci::Error),
+
+    /// Failed to get vfio host info
+    #[error("PCI get host info failed: {0}")]
+    GetHostInfo(String),
+
+    /// Invalid PCI device ID
+    #[error("invalid PCI device ID: {0}")]
+    InvalidDeviceID(u32),
+
+    /// Failed to allocate device resource
+    #[error("failure while allocate device resource: {0:?}")]
+    AllocateDeviceResource(#[source] ResourceError),
+
+    /// Vfio container not found
+    #[error("vfio container not found")]
+    VfioContainerNotFound,
+
+    /// Generic IO error.
+    #[error("Generic IO error, {0}")]
+    IoError(#[source] std::io::Error),
+}
+
+type Result<T> = std::result::Result<T, VfioDeviceError>;
+
+/// Host info for vfio device
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct VfioDeviceHostInfo {
+    pub group_id: u32,
+    pub group_fd: RawFd,
+    pub device_fd: RawFd,
+}
+
+/// Configuration information for a VFIO PCI device.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize, Default)]
+pub struct VfioPciDeviceConfig {
+    /// PCI device information: "bus:slot:function"
+    pub bus_slot_func: String,
+    /// PCI vendor and device id
+    /// high 16bit : low 16bit = device_id : vendor_id
+    pub vendor_device_id: u32,
+    /// Deice ID used in guest, guest_dev_id = slot
+    pub guest_dev_id: Option<u8>,
+    /// Clique ID for Nvidia GPUs and RDMA NICs
+    pub clique_id: Option<u8>,
+}
+
+impl VfioPciDeviceConfig {
+    ///default pci domain is 0
+    pub fn host_pci_domain(&self) -> u32 {
+        0
+    }
+    pub fn valid_vendor_device(&self) -> bool {
+        if self.vendor_device_id == 0 {
+            return true;
+        }
+        self.vendor_device_id != 0
+            && (self.vendor_device_id & 0xffff) != 0
+            && ((self.vendor_device_id >> 16) & 0xffff) != 0
```

Comment:
Add two consts here for `16` and `0xffff`, one named to `OFFSET`, one named to `MASK`.

Reply:
document add here.